### DISCUSSION TITLE
Is it possible to refresh data without deleting the source CSV

### DISCUSSION THREAD

**Discussion by ezvr** [binjr/binjr#153] [Q&A] [answered]
I have a csv file that gets updated by a script. Currently, I remove the CSV from the Sources panel, then re-add it. However this means I also have to re-create the graphs every single time.

Is it possible to make binjr just refresh the data somehow?

**Comment by Arkanosis** [comment]
Hi @ezvr 

I don't think there's (yet) a refresh feature for a CSV source. However, to avoid having to re-create many graphs every time, you could save the workspace, and open it again to get the new values in the old graphs. Assuming you've already saved the workspace, re-opening is a two-clicks operation.

![steps](https://github.com/user-attachments/assets/3ebf299b-6954-4d82-be41-9d6710910b46)

Best regards,

**Reply by ezvr** [reply]
thank you! that is a good workaround

**Comment by fthevenet** [comment]
Related https://github.com/binjr/binjr/issues/96
Need to look into that again, sometimes.

**Comment by fthevenet** [comment, marked as answer]
As a matter of fact, the log files plugin already supports reloading all the data from the underlying file upon doing a "hard refresh" (`Ctrl+F5` or `Ctrl`+ left mouse click on the refresh button), which basically does what you are asking.
It seemed quite reasonable to extend the CSV plugin so that it could do the same, so I went ahead and did just that: you can try it in this preview build: https://github.com/binjr/binjr/releases/tag/v3.20.0-SNAPSHOT

Please be aware that there are limitations to what it can do:
* Unlike what is proposed in #96, this will take into account new rows or modified/deleted rows in CSV, but if a new column is added, it will not appear in the source pane until you close and reopen it.
* Doing a hard refresh causes a complete reloading and re-index of the whole file, which means it is not an efficient way to capture very frequent changes on large files.
* Once you have done a hard refresh, binjr will be aware of data that were newly added to a CSV file, but remember that the time range for an existing worksheet will not have changed. So after pressing `Ctrl+F5` to refresh the data, you may need to click the "Reset Time Range" button (or press `F9`) for the charts to show the new data.

**Reply by fthevenet** [reply]
@Arkanosis Please be aware that I had to change parts of the Worksheet API to make them more generic, so this _may_ affect downstream plugins: you might want to test early that it doesn't affect yours.

**Reply by Arkanosis** [reply]
> I had to change parts of the Worksheet API to make them more generic

Thanks a lot for the heads up! I'll look at that tomorrow :)

**Comment by apiszcz** [comment]
Automatic refresh of CSV contents on an interval settable by the user would be great.
Outstanding work on this capability!

**Reply by fthevenet** [reply]
Hi, thanks for your interest.

Live monitoring is currently an explicit non-goal for binjr (as stated in the [README](https://github.com/binjr/binjr?tab=readme-ov-file#what-is-binjr)); not because it's a bad idea, but simply because **doing it right** would require more work than I can commit to right now (i.e. it would need to work across all source types, automatically adapt the displayed time range, and for CSV files, be more efficient than just re-indexing the whole file, etc...)

Cheers!

**Comment by apiszcz** [comment]
understand.
Thank you. Time is limited!
